executor/join.go

```diff
@@ -665,46 +650,99 @@ func (e *HashJoinExec) handleFetchInnerAndBuildHashTablePanic(r interface{}) {
 }
 
 func (e *HashJoinExec) fetchInnerAndBuildHashTable(ctx context.Context) {
-	// innerResultCh transfers inner chunk from inner fetch to build hash table.
-	innerResultCh := make(chan *chunk.Chunk, 1)
-	doneCh := make(chan struct{})
-	// TODO: just used in test now, thus we return directly in this if-branch.
-	// Remained work:
-	// 1. parallel build hash tables for the inner sub-partitions
-	// 2. partition the outer relation into sub-partitions
-	// 3. parallel join corresponded inner-partitions with outer-partitions
+	// Step1: fetch all the data of the inner relation into memory.
+	if err := e.fetchInnerRows(ctx); err != nil {
+		e.innerFinished <- err
+		return
+	}
+	if e.finished.Load().(bool) {
+		return
+	}
+
+	// Step2: do parallel partition if needed.
+	needPartition := false
 	if e.ctx.GetSessionVars().EnableRadixJoin {
-		go util.WithRecovery(func() {
-			e.fetchInnerRows(ctx, innerResultCh, doneCh)
-			if !e.evalRadixBit() {
-				return
-			}
-			if e.partitionInnerRows() != nil {
+		needPartition = e.evalRadixBit()
+		if needPartition {
+			if err := e.partitionInnerRows(); err != nil {
+				e.innerFinished <- err
 				return
 			}
-
-		}, nil)
+		}
+	}
+	if e.finished.Load().(bool) {
 		return
 	}
 
-	go util.WithRecovery(func() { e.fetchInnerRows(ctx, innerResultCh, doneCh) }, nil)
+	// Step3: build hash table(s) for the inner sub-relation.
+	if err := e.buildHashTable(needPartition); err != nil {
+		e.innerFinished <- err
+	}
+}
 
-	// TODO: Parallel build hash table. Currently not support because `mvmap` is not thread-safe.
-	err := e.buildHashTableForList(innerResultCh)
-	if err != nil {
-		e.innerFinished <- errors.Trace(err)
-		close(doneCh)
+func (e *HashJoinExec) buildHashTable(hasPartitionPhase bool) error {
+	e.innerKeyColIdx = make([]int, len(e.innerKeys))
+	for i := range e.innerKeys {
+		e.innerKeyColIdx[i] = e.innerKeys[i].Index
+	}
+
+	if !hasPartitionPhase {
+		return e.buildGlobalHashTable()
 	}
-	// wait fetchInnerRows be finished.
-	for range innerResultCh {
+	return e.buildHashTable4Partitions()
+}
+
+func (e *HashJoinExec) wait4BuildHashTable(wg *sync.WaitGroup, finishedCh chan error) {
+	wg.Wait()
+	close(finishedCh)
+}
+
+func (e *HashJoinExec) buildHashTable4Partitions() error {
+	e.hashTables = make([]*mvmap.MVMap, len(e.innerParts))
+	buildFinishedCh := make(chan error, e.concurrency)
+	wg := &sync.WaitGroup{}
+	wg.Add(int(e.concurrency))
+	go e.wait4BuildHashTable(wg, buildFinishedCh)
+	for i := 0; i < int(e.concurrency); i++ {
+		workerID := i
+		go util.WithRecovery(func() {
+			defer wg.Done()
+			e.doBuild(workerID, buildFinishedCh)
+		}, nil)
 	}
+	return <-buildFinishedCh
```

Comment:
This control goroutine `go e.wait4BuildHashTable(wg, buildFinishedCh)` can be removed?
Move its code to the end of buildHashTable4Partitions is better?
```suggestion
wg.Wait()
close(buildFinishedCh)
return <-buildFinishedCh
```

Reply:
wg.Wait will wait for the end of all the workers, but we want the error can be returned as soon as possible.